py/emitnative: Further Viper integer-indexed load/store improvements.

agatti opened the pull request:
### Summary

This PR introduces improvements in the code that gets generated for integer-indexed load and store operations for various architectures.

The Viper integer-indexed generator bits are now able to utilise a specialised operation implementation if the platform emitter exposes one.  This means that macros named `ASM_{LOAD,STORE}{8,16,32}_REG_REG_OFFSET` are used if they are present in the platform backend.

`ASM_LOAD16_REG_REG_OFFSET`, `ASM_LOAD32_REG_REG_OFFSET`, and `ASM_STORE32_REG_REG_OFFSET` were already present since they are used by the native emitter, so their implementations are used also by the Viper emitter now.  This may or may not reduce the code size, but it streamlines the current behaviour by having one single optimised implementation on each platform.

However, those three implementations are now exposed to scenarios where accessing larger than usual offsets is a somewhat probable occurrence - requiring some changes to make them work in all possible cases on Arm, Thumb, and Xtensa.  The full set of platform-optimised integer indexed load/store operations is not implemented for all Viper operations on all platforms, as the focus of this PR is ARMv7-M improvements.  The rest of the changes is there to adapt the other platform backends to the infrastructure changes needed for ARMv7-M.  The remaining load/store operations' code will come later in a different PR if the code size increase can be kept to a minimum.

The Arm emitter is now able to generate `LDR` and `STR` integer-indexed opcodes whose offsets can span the full 32 bits address space.  The original behaviour would be undefined if the offset was larger than 12 bits, as it would corrupt the upper bits of the generated opcode itself.

The Thumb emitter is now able to generate ARMv7-M opcodes for all data sizes if both ARMv7-M opcodes are allowed and if the opcode offset spans between 6 and 12 bits.  In that case a single 32 bits opcode will be generated (`LDR{B,H}.W` and `STR{B,H}.W`), falling back to the existing generic code sequence generator if the offset is too big to be embedded in a single opcode.

The Xtensa emitter is now able to generate opcode sequences spanning the full 32 bits address space for 32 bits loads and stores, and for 16 bit loads.  The existing implementations would raise a `RuntimeError` if the offset could not be embedded in a single opcode.

### Testing

The native and viper tests in `tests/micropython` ran successfully on QEMU `MPS2-AN385`, `SABRELITE`, and `VIRT_RV32` boards, on a NodeMCU v3 8266 board, on a generic ESP32 D0WD board, and on x64 via the Unix port.

### Trade-offs and Alternatives

I've tried to keep the extra code footprint as limited as possible without having to make the code too complex to follow (leaving some potential ARMv7-M only optimisations out as well, taking up a bit more space).

This is mostly evident for armv6/armv7 targets, where the overhead should be ~200 bytes.  This was offset a bit by moving out some code from the native emitter to the platform backends, where a simpler implementation could suffice.

However there are some benefits in more optimised code being generated, and in a slight decrease of complexity in the Viper emitter as less platform-specific code is present there.

agatti committed:
py/emitnative: Let Viper int-indexed code use appropriate operands.

This commit extends the generic ASM API by adding the rest of the
ASM_{LOAD,STORE}[size]_REG_REG_OFFSET macros whenever applicable.

The Viper int-indexed load/store code generator was changed to use those
API functions if they are available, falling back to backend-specific
implementations if possible and ultimately to a generic implementation.

Right now all backends except for x64 implement load16, load32, and
store32 operations (x64 only implements load16).

Signed-off-by: Alessandro Gatti <a.gatti@frob.it>

agatti committed:
py/asmthumb: Extend load/store generators with ARMv7-M opcodes.

This commit lets the Thumb native code generator backend emit ARMv7-M
specific opcodes for indexed load/store operations if possible.

Now T3 opcode encodings are used if the generator backend is configured
to allow emitting ARMv7-M opcodes and if the (unsigned) scaled index
fits in 12 bits.  Or, in other words, LDR{B,H}.W and STR{B,H}.W opcodes
are now emitted if possible.

Signed-off-by: Alessandro Gatti <a.gatti@frob.it>

agatti committed:
py/asmxtensa: Extend existing specialised load/store operations range.

This commit updates the existing specialised implementations for
int-indexed 32-bit load and store operations, and adds a specialised
implementation for int-indexed 16-bit load.

The 32-bit operations relied on the fact that their applicability was
limited to a specific range, falling back on a generic implementation
otherwise.  Introducing a single entry point for each int-indexed
load/store operation size would break that assumption.  Now those two
operations contain fallback code to generate working code by themselves
instead of raising an exception.

The 16-bit operation instead simply did not have any range check, but it
was not exposed directly to the Viper emitter.  When a 16-bit
int-indexed load entry point was introduced, the existing implementation
would fail when accessing memory outside its 0..255 halfwords range.  A
specialised implementation is now present, performing fewer operations
than the existing Viper emitter equivalent.

Signed-off-by: Alessandro Gatti <a.gatti@frob.it>

agatti committed:
py/emitnative: Remove redundant RV32 Viper int-indexed code.

This commit removes redundant RV32 implementations of certain
int-indexed code generation operations (32-bit load/store and 16-bit
load).

Those operations were already available as part of the native emitter
API but were not exposed to the Viper code generator.  As part of the
introduction of more specialised load and store API calls to
int-indexed Viper load/store generator bits, the existing native emitter
implementations are reused, thus making the Viper implementations
redundant.

Signed-off-by: Alessandro Gatti <a.gatti@frob.it>

agatti committed:
py/asmarm: Extend int-indexed 32-bit load/store offset ranges.

This commit extends the range for int-indexed load/store opcode
generators, making them emit correct code sequences for offsets that
span more than 12 bits.

This is necessary due to those generator bits being also used in the
Viper emitter, where it's more probable to reference offsets that can
not be embedded in the LDR/STR opcodes.

Signed-off-by: Alessandro Gatti <a.gatti@frob.it>

agatti committed:
tests/run-tests.py: Unconditionally enable native tests if asked.

This commit lets the test runner enumerate and run native tests if the
feature check fails but native tests were explicitly requested from the
command line.

The old behaviour would disable native tests anyway if the feature check
failed, however this hid a bug in the x86 native emitter that would be
triggered even during the feature check.  That meant the test suite
would pass on x86 even with a broken emitter, as those tests would have
been skipped anyway.

Now, if the user asks for native code it will get native code out of the
runner no matter what.

Co-authored-by: Damien George <damien@micropython.org>
Signed-off-by: Alessandro Gatti <a.gatti@frob.it>

agatti committed:
tools/ci.sh: Fix nanbox CI test runs.

This commit fixes CI test runs for the `nanbox` target, which were
broken by the unconditional native emitter code output changes in the
test runner.

The `nanbox` configuration does not enable native emitters of any kind,
and with a full test run that includes executing emitted native code
things would break when doing CI runs.

This is worked around by introducing a common subset of tests that do
not involve the native emitter, and a more comprehensive set of tests
that include both non-emitter and emitter tests.  The `nanbox` CI test
run will stop at the first subset, whilst other configurations will run
that and execute further tests.

Function names have been kept the same for steps that involve native
code, with the `nanbox` subset having another one.  This should not
trigger any breakage in existing CI configurations or external scripts.

Signed-off-by: Alessandro Gatti <a.gatti@frob.it>

agatti commented:
There are lots of ways to work around the `nanbox` CI issue - please let me know if the one I chose is OK with you.

dpgeorge commented on py/asmxtensa.c:
```diff
@@ -42,6 +42,10 @@
 
+// A7 is part of the incoming parameter registers set in both CALL0 and
+// windowed ABIs, thus is not callee-saved and can be safely clobbered.
+#define REG_TEMP ASM_XTENSA_REG_A7
```
This is problematic: A7 is used as a base pointer to the function table `mp_fun_table` when in windowed mode.  So clobbering it will lead to a crash if the function table is used afterwards.

Eg the following crashes on esp32:
```py
@micropython.native
def f(p: ptr32):
    p[256] = 1
    print('done')

f(bytearray(2048))
```
It's a bit tricky to fix.  The native emitter does allocate a register for calculating the index `reg_index` but that's not passed through to the backend emitter.

agatti replied:
This is weird, as I'm quite sure I ran the full set of `tests/micropython/` tests on a D0WD.  `tests/micropython/viper_ptr32_store_boundary.py` would have crashed when using the last set of offsets (around the 12-bits boundary mark).

But I've seen the constant in question so this needs fixing.  

I believe that using `REG_TEMP2` could be a temporary workaround - don't have an ESP32 available at the moment to test it.

`REG_TEMP2` is only used for implementing the `rot_three` opcode, and the native implementation whilst performing register stores would only use the stack for with that register.  How big can a single emitted function stack grow?  If that number is below 1024 bytes then `REG_TEMP2` should be it, otherwise I'll see how this can be worked around differently.

dpgeorge replied:
> `tests/micropython/viper_ptr32_store_boundary.py` would have crashed when using the last set of offsets (around the 12-bits boundary mark).

The reason it doesn't crash is because it doesn't access A7 after clobbering it.  In my test above you need the `print('done')` to trigger the crash, because that line requires A7.

agatti replied:
OK, using `REG_TEMP2` (or the same register anyway) seems to work on a regular ESP32 board when running the test suite with `--via-mpy --emit native`.

I've had to extend the range of some branch opcodes to make `basics/frozenset_binop.py`, `basics/set_binop.py`, `extmod/framebuf1.py`, and `extmod/json_dumps_separators.py` actually compile for xtensa/xtensawin in the first place.

I haven't tested this on an ESP8266 yet (need to fetch the board from somewhere else) but, unless checking for `N_XTENSAWIN` when picking which register number to use is not correct, in this case if it works on ESP32 then it should also work for ESP8266.

dpgeorge replied:
> I've had to extend the range of some branch opcodes to make `basics/frozenset_binop.py`, `basics/set_binop.py`, `extmod/framebuf1.py`, and `extmod/json_dumps_separators.py` actually compile for xtensa/xtensawin in the first place.

Excellent, thanks for that.  I had noticed for some time that these tests fail, but with all the other native changes in flight, I was waiting before mentioning it.  See related #17293.

agatti replied:
> I haven't tested this on an ESP8266 yet (need to fetch the board from somewhere else) but, unless checking for `N_XTENSAWIN` when picking which register number to use is not correct, in this case if it works on ESP32 then it should also work for ESP8266 unless this throws alignment off for inlineasm tests if I didn't handle things right when updating branch opcodes.

I've just had the chance to test this on an ESP8266 - inline asm branches aren't affected, some test crash with `--via-mpy --emit=native`.  I'm looking into it.

agatti replied:
OK, I've misread the ISA document - `BEQZ.N` and `BNEZ.N` can only operate with an *unsigned* displacement, and on an ESP32 there were no backward jumps below the (signed) 6-bits threshold that triggered the emission of those opcodes.

This should be fixed now!

dpgeorge commented on py/asmthumb.h:
```diff
@@ -462,16 +464,21 @@ void asm_thumb_b_rel12(asm_thumb_t *as, int rel);
-#define ASM_STORE_REG_REG_OFFSET(as, reg_src, reg_base, word_offset) asm_thumb_str_rlo_rlo_i5((as), (reg_src), (reg_base), (word_offset))
-#define ASM_STORE8_REG_REG(as, reg_src, reg_base) asm_thumb_strb_rlo_rlo_i5((as), (reg_src), (reg_base), 0)
-#define ASM_STORE16_REG_REG(as, reg_src, reg_base) asm_thumb_strh_rlo_rlo_i5((as), (reg_src), (reg_base), 0)
-#define ASM_STORE32_REG_REG(as, reg_src, reg_base) asm_thumb_str_rlo_rlo_i5((as), (reg_src), (reg_base), 0)
```
These inline functions like `asm_thumb_strb_rlo_rlo_i5()` are now unused, so could be removed.  But maybe worth keeping them around in case they are needed again one day?

agatti replied:
They're not adding any code to the final binary and they're under version control anyway.  I'd personally take them out but maybe some external code is relying on those functions?

dpgeorge replied:
Let's leave them in for now, until all these changes to the native emitter settle down.

dpgeorge commented on py/asmarm.c:
```diff
@@ -333,9 +333,16 @@ void asm_arm_asr_reg_reg(asm_arm_t *as, uint rd, uint rs) {
+    } else {
+        // mov r8, #off
+        // ldr rd, [rn, r8]
+        asm_arm_mov_reg_i32_optimised(as, ASM_ARM_REG_R8, byte_offset);
```
Is it worth doing:
```c
#define REG_TEMP ASM_ARM_REG_R8
```
and then using `REG_TEMP` here?  And also change the other locations in this file that use r8.  Then at least it's clear that this is an available temporary register.

agatti replied:
I briefly thought about it when touching the code but I decided to leave things alone to minimise the amount of code touched by that particular commit - same thing for all the checks for `(reg < ASM_THUMB_REG_R8)` in `py/asmthumb.c` that can be moved into a macro.

Anyway, I've added a commit to do just that, no big deal.

agatti committed:
py/asmxtensa: Extend BCCZ range to 18 bits.

This commit lets the native emitter backend extends the range of the
BCCZ family of opcodes (BEQZ, BNEZ, BLTZ, BGEZ) from 12 bits to 18
bits.

The test suite contains some test files that, when compiled into native
code, would require BCCZ jumps outside the (signed) 12 bits range.  In
this case either the MicroPython interpreter or mpy-cross would raise an
exception, not running the test when using the "--via-mpy --emit native"
command line options with the test runner.

This comes with a 3 bytes penalty on each forward jump, bringing the
footprint of those jumps to 6 bytes each, as a longer opcode sequence
has to be emitted to let jumps access a larger range.  However, this is
slightly offset by the fact that backward jumps can be emitted with a
single opcode if the range is small enough (3 bytes for a 12-bits
offset).

Signed-off-by: Alessandro Gatti <a.gatti@frob.it>

agatti committed:
py/asmxtensa: Extend BCC range to 18 bits.

This commit lets the native emitter backend extends the range of the
BCC family of opcodes (BALL, BANY, BBC, BBS, BEQ, BGE, BGEU, BLT,
BLTU, BNALL, BNE, BNONE) from 8 bits to 18 bits.

The test suite contains some test files that, when compiled into native
code, would require BCC jumps outside the (signed) 8 bits range.  In
this case either the MicroPython interpreter or mpy-cross would raise an
exception, not running the test when using the "--via-mpy --emit native"
command line options with the test runner.

This comes with a 3 bytes penalty on each forward jump, bringing the
footprint of those jumps to 6 bytes each, as a longer opcode sequence
has to be emitted to let jumps access a larger range.  However, this is
slightly offset by the fact that backward jumps can be emitted with a
single opcode if the range is small enough (8-bits offset).

Signed-off-by: Alessandro Gatti <a.gatti@frob.it>

agatti committed:
py/asmarm: Give a proper name to the temporary register.

This commit performs a small refactoring on the Arm native emitter, by
renaming all but one instance of ASM_ARM_REG_R8 into REG_TEMP.

ASM_ARM_REG_R8 is the temporary register used by the emitter when
operations cannot overwrite the value of a particular register and some
extra storage is needed.

Signed-off-by: Alessandro Gatti <a.gatti@frob.it>

dpgeorge reviewed (approved):
Thanks for updating/fixing.  It looks like everything is resolved now.

Tested on ESP8266_GENERIC and ESP32_GENERIC, with `run-tests.py --via-mpy --emit native`.  Not everything passes, but those that fail did already fail before this PR.
